### PR TITLE
Implement all_gather with native XLA primitive.

### PR DESCRIPTION
**Pull request by hjm-aws**
Hi,

I wasn't able to test this on CPU or TPU. I believe all-gather is not lowered on CPU yet, and I am not sure how to acquire a TPU instance to run this. If you can point me to the instructions it will be much appreciated!

For issue #3138

### CODE REVIEW

**Comment by hjm-aws** [comment]
The build failure is due to pytorch commit https://github.com/pytorch/pytorch/commit/36db501736cafd69cdfd0114b6136b6cc2e91160#diff-2f3dbd85efb9b5172f2264eedd3be47dd765e6ab7cc8bf3ade5e62c28ae35991L9048. I am not sure how to make my PR depend on an earlier pytorch commit.

**Comment by JackCaoG** [comment]
@hjm-aws FYI, adding a pin like https://github.com/pytorch/xla/pull/3208/files#diff-e71f55b9deb410fad9d5945c4c30878a93369434520520987ac03c59bfa5280d can help you ping the upstream pytorch pr on circle CI

**Review by JackCaoG — COMMENTED** [review]
Mostly LGTM, want to get some clarification on the test change.

File: test/test_mp_all_gather.py [hunk 1]
```diff
@@ -20,8 +21,7 @@ def _mp_fn(index):
       sys.exit(1)
   else:
     print(
-        'Default device {} does not support replication'.format(device),
-        file=sys.stderr)
+        'Default device {} is not a TPU device'.format(device), file=sys.stderr)
```
[the remark this answers is not in the document]
Comment (JackCaoG) [reply test/test_mp_all_gather.py:24]: If you rebase this pr, circle CI should build and run the test on cpu and gpu. I think gpu should work.

**Comment by JackCaoG** [comment]
@hjm-aws Can you rebase this pr and remove the `TPU` hard check? Circle CI should run this time for CPU and GPU.

File: test/test_mp_all_gather.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hjm-aws) [reply test/test_mp_all_gather.py:24]: Done. Sorry for the late response!

**Comment by JackCaoG** [comment]
@hjm-aws Thanks. The commit history seems a bit messed up, it bring into commits that already merged to the master.

**Comment by hjm-aws** [comment]
@JackCaoG Yes, I thought it was normal (this is the first PR I created). What I did was the following:
1. Create a fork https://github.com/hjm-aws/xla/tree/master.
2. Create a branch on the fork https://github.com/hjm-aws/xla/tree/all_gather.
3. Use https://github.com/pytorch/xla as an upstream remote.
4. Created this PR in https://github.com/hjm-aws/xla/tree/all_gather.
5. After seeing you rebase request, firstly in https://github.com/hjm-aws/xla/tree/master, I did `git pull upstream master`. Then in https://github.com/hjm-aws/xla/tree/all_gather, I did `git rebase master`, `git pull --rebase`, and `git push`.

What is the proper way to rebase this PR?

**Comment by hjm-aws** [comment]
@JackCaoG OK, seems I cleaned up the commit history. The only garbage left is in the conversation history.

**Review by JackCaoG — APPROVED** [review]
Thanks @hjm-aws !

**Pull request merged**